README.md
# Audio record streamlit

This streamlit component allows to register an audio utterence from a user.

<img src="https://raw.githubusercontent.com/Joooohan/audio-recorder-streamlit/main/img/recorder.png" alt="recorder.png" width="200"/>

## Installation

`pip install audio-recorder-streamlit`

## Usage

```python
import streamlit as st
from audio_recorder_streamlit import audio_recorder

audio_bytes = audio_recorder()
if audio_bytes:
    st.audio(audio_bytes, format="audio/wav")
```

## Recording parameters

You can adjust the recording parameters `energy_threshold` and
`pause_threshold`:
- `energy_threshold`: The energy recording sensibility above which we consider
    that the user is speaking. If it is a float, then this is the energy
    threshold used to automatically detect recording start and recording end.
    You can provide a tuple for specifying different threshold for recording
    start detection and recording end detection.
- `pause_threshold`: The number of seconds to spend below `energy_level` to
    automatically stop the recording.
- `sample_rate`: Sample rate of the recorded audio. If not provided, this will use the
    default sample rate (https://developer.mozilla.org/en-US/docs/Web/API/AudioContext/AudioContext).


```python
# The recording will stop automatically
# 2 sec after the utterance end
audio_bytes = audio_recorder(pause_threshold=2.0, sample_rate=41_000)
```

## Styling parameters

You can adjust the button style parameters:
- **Text**: specify the text displayed next to the icon
- **Icon**: specify the icon among the font awesome solid icons
  (https://fontawesome.com/search?o=r&s=solid)
- **Color**: specify the neutral color and recording color.
- **Size**: specify the icon size using font awesome sizing
  (https://fontawesome.com/docs/web/style/size).

```python
audio_bytes = audio_recorder(
    text="",
    recording_color="#e8b62c",
    neutral_color="#6aa36f",
    icon_name="user",
    icon_size="6x",
)
```
<img src="https://raw.githubusercontent.com/Joooohan/audio-recorder-streamlit/main/img/custom.png" alt="custom.png" width="200"/>

# Frequently Asked Question
## How can I record for a fixed duration ?

You can record for a fixed duration by setting the `energy_threshold=(-1.0, 1.0)`
so that the recorder considers that you are speaking at the beginning and then
you are never speaking from this point on.

Then simply set `pause_threshold` to your desired recording length.

```python
# Records 3 seconds in any case
audio_bytes = audio_recorder(
  energy_threshold=(-1.0, 1.0),
  pause_threshold=3.0,
)
```


CHANGELOG.md
# Changelog

# v0.0.8 (24/01/2023)
- Added the `sample_rate` parameter.
- Updated the `energy_threshold` parameter to be a `float` or a `(float,
  float)`. This allows to parameter a fixed length recording.


tests/test_sanity.py
from audio_recorder_streamlit import _RELEASE

assert _RELEASE


audio_recorder_streamlit/__init__.py
import json
import os
from typing import Optional

import streamlit.components.v1 as components

_RELEASE = True

if _RELEASE:
    parent_dir = os.path.dirname(os.path.abspath(__file__))
    build_dir = os.path.join(parent_dir, "frontend/build")
    _audio_recorder = components.declare_component(
        "audio_recorder", path=build_dir
    )
else:
    _audio_recorder = components.declare_component(
        "audio_recorder",
        url="http://localhost:3001",
    )


def audio_recorder(
    text: str = "Click to record",
    energy_threshold: float = 0.01,
    pause_threshold: float = 0.8,
    neutral_color: str = "#303030",
    recording_color: str = "#de1212",
    icon_name: str = "microphone",
    icon_size: str = "3x",
    sample_rate: Optional[int] = None,
    key: Optional[str] = None,
) -> Optional[bytes]:
    """Create a new instance of "audio_recorder".

    Parameters
    ----------
    text: str
        The text to display next to the recording button.
    energy_threshold: Union[float, Tuple[float, float]]
        The energy recording sensibility above which we consider that the user
        is speaking. If it is a float, then this is the energy threshold used
        to automatically detect recording start and recording end. You can
        provide a tuple for specifying different threshold for recording start
        detection and recording end detection.
    pause_threshold: float
        The number of seconds to spend below `energy_level` to automatically
        stop the recording.
    neutral_color: str
        Color of the recorder icon while stopped.
    recording_color: str
        Color of the recorder icon while recording.
    icon_name: str
        Font Awesome solid icon name
        (https://fontawesome.com/search?o=r&s=solid)
    icon_size: str
        Size of the icon (https://fontawesome.com/docs/web/style/size)
    sample_rate: Optional[int]
        Sample rate of the recorded audio. If not provided, this will use the
        default sample rate
        (https://developer.mozilla.org/en-US/docs/Web/API/AudioContext/AudioContext).
    key: str or None
        An optional key that uniquely identifies this component. If this is
        None, and the component's arguments are changed, the component will be
        re-mounted in the Streamlit frontend and lose its current state.

    Returns
    -------
    Optional[bytes]
        Bytes representing the recorded audio in the `audio/wav` format.

    """
    if type(energy_threshold) in [list, tuple]:
        start_threshold, end_threshold = energy_threshold
    else:
        start_threshold = energy_threshold
        end_threshold = energy_threshold

    data = _audio_recorder(
        text=text,
        start_threshold=start_threshold,
        end_threshold=end_threshold,
        pause_threshold=pause_threshold,
        neutral_color=neutral_color,
        recording_color=recording_color,
        icon_name=icon_name,
        icon_size=icon_size,
        sample_rate=sample_rate,
        key=key,
        default=None,
    )
    audio_bytes = bytes(json.loads(data)) if data else None
    return audio_bytes


if not _RELEASE:
    import streamlit as st

    st.subheader("Base audio recorder")
    base_audio_bytes = audio_recorder(key="base")
    if base_audio_bytes:
        st.audio(base_audio_bytes, format="audio/wav")

    st.subheader("Custom recorder")
    custom_audio_bytes = audio_recorder(
        text="",
        recording_color="#e8b62c",
        neutral_color="#6aa36f",
        icon_name="user",
        icon_size="6x",
        sample_rate=41_000,
        key="custom"
    )
    st.text("Click to record")
    if custom_audio_bytes:
        st.audio(custom_audio_bytes, format="audio/wav")

    st.subheader("Fixed length recorder")
    fixed_audio_bytes = audio_recorder(
        energy_threshold=(-1.0, 1.0),
        pause_threshold=3.0,
        key="fixed",
    )
    st.text("Click to record 3 seconds")
    if fixed_audio_bytes:
        st.audio(fixed_audio_bytes, format="audio/wav")


audio-recorder/frontend/src/RecordSection.js
import React, { useState, useEffect } from "react";
import tw from "twin.macro";
// import styled from "styled-components";
import { SectionHeading as HeadingTitle } from "../misc/Headings.js";
import { ReactComponent as SvgDecoratorBlob1 } from "../../images/svg-decorator-blob-1.svg";
import { ReactComponent as SvgDecoratorBlob2 } from "../../images/svg-decorator-blob-3.svg";
import { ReactComponent as SvgPrevIcon } from "../../images/prev-icon.svg";
import { ReactComponent as SvgNextIcon } from "../../images/next-icon.svg";
import AudioReactRecorder, {
  RecordState,
} from "components/accentpal/AudioRecorder";
import { config } from "Constants.js";
import RecordButton from "components/accentpal/RecordButton.js";
import PlayButton from "components/accentpal/PlayButton.js";
import PhonemeViewer from "components/accentpal/PhonemeViewer.js";
import { Tooltip } from "@material-ui/core";

const Container = tw.div`relative`;
const Content = tw.div`max-w-screen-xl mx-auto py-20 lg:py-24`;
const HorizontalContainer = tw.div`flex flex-row items-center`;
const VerticalContainer = tw.div`flex flex-col items-center`;

const HeadingDescription = tw.p`mt-2 text-xl font-medium text-gray-600 text-center max-w-sm`;
const TextToRead = tw.p`m-4 font-normal text-gray-800 text-center text-xl max-w-sm`;
const Audio = tw.audio`mt-4`;
const Image = tw.img`m-4 h-40 w-48 sm:h-64 sm:w-96 rounded-lg inline`;

const DecoratorBlob1 = tw(
  SvgDecoratorBlob1
)`-z-10 absolute bottom-0 right-0 w-48 h-48 transform translate-x-40 -translate-y-8 opacity-25`;
const DecoratorBlob2 = tw(
  SvgDecoratorBlob2
)`-z-10 absolute top-0 left-0 w-48 h-48 transform -translate-x-32 translate-y-full opacity-25`;

export default () => {
  const title = "Try the demo";
  const description = "Click and speak";

  const [recordState, setRecordState] = useState(null);
  const [audioData, setAudioData] = useState(null);
  const [text, setText] = useState("no word");
  const [imageSrc, setImageSrc] = useState("");
  const [target, setTarget] = useState([]);
  const [pred, setPred] = useState([]);
  const [recorderHeight, setRecorderHeight] = useState("150");
  const [examples, setExamples] = useState([]);
  const [examplesPhonemes, setExamplesPhonemes] = useState([]);
  const [current, setCurrent] = useState(0);
  const [recordColor, setRecordColor] = useState("white");
  const [radius, setRadius] = useState("0");
  const [playCircleRef, setPlayCircleRef] = useState("#F0F3F4");
  const [playIconRef, setPlayIconRef] = useState("#4F4F4F");
  const [playCircleSelf, setPlayCircleSelf] = useState("#F0F3F4");
  const [playIconSelf, setPlayIconSelf] = useState("#4F4F4F");
  const _cache = {};

  useEffect(() => {
    async function submitAudio() {
      if (!audioData) {
        return;
      }

      const fdata = new FormData();
      fdata.append("text", text);
      fdata.append("audio", audioData.blob);
      await fetch(`${config.API_URL}/api/utterance`, {
        method: "POST",
        body: fdata,
      })
        .then((response) => response.json())
        .then((data) => presentResult(data))
        .catch((error) =>
          console.log(`An error occurred sending the audio. ${error}`)
        );
    }

    submitAudio();
  }, [audioData, text]);

  useEffect(() => {
    async function fetchSentence() {
      await fetch(`${config.API_URL}/api/sentence`, {
        method: "GET",
      })
        .then((response) => response.json())
        .catch((error) =>
          console.log(` Error: fetching the sentence to pronounce. ${error}`)
        )
        .then((data) => {
          const fetched_examples = data["examples"];
          setExamples(fetched_examples);
          setText(fetched_examples[current]["word"]);
          setImageSrc(fetched_examples[current]["imageSrc"]);
          setTarget(fetched_examples[current]["target"]);
        })
        .catch((error) =>
          console.log(
            `An error occured while fetching the sentence to pronounce. ${error}`
          )
        );
    }

    fetchSentence();
  }, []);

  useEffect(() => {
    async function fetchPhonemesExamples() {
      await fetch(`${config.API_URL}/api/examples`, {
        method: "GET",
      })
        .then((response) => response.json())
        .then((data) => {
          setExamplesPhonemes(data["examples"]);
          console.log(data["examples"]);
        })
        .catch((error) =>
          console.log(
            `An error occured while fetching the phonemes examples. ${error}`
          )
        );
    }

    fetchPhonemesExamples();
  }, []);

  useEffect(() => {
    if (examples.length > 0) {
      setText(examples[current]["word"]);
      setImageSrc(examples[current]["imageSrc"]);
      setTarget(examples[current]["target"]);
      setAudioData(null);
      setPred("");
    }
  }, [current, examples]);

  const presentResult = (data) => {
    setPred(data["pred"]);
  };

  const handleClick = () => {
    if (recordState !== RecordState.START) {
      setRecordState(RecordState.START);
      setRecorderHeight("0");
      setTimeout(() => {
        setRecordColor("#FBC4AC");
      }, 500);
    } else {
      setRecordState(RecordState.STOP);
    }
  };

  const handleStop = (data) => {
    setRecordState(RecordState.STOP);
    setRecorderHeight("0");
    setRadius("0");
    setRecordColor("white");
    setAudioData(data);
  };

  const getVoices = (locale) => {
    if (_cache[locale]) return _cache[locale];

    const synth = window.speechSynthesis;
    const _voices = synth.getVoices();
    _cache[locale] = _voices.filter((voice) => voice.lang === locale);
    return _cache[locale];
  };

  const speakWord = () => {
    const synth = window.speechSynthesis;
    const utterThis = new SpeechSynthesisUtterance();
    const voices = getVoices("en-US");
    utterThis.voice = voices[0];
    utterThis.lang = "en-US";
    utterThis.rate = 0.7;
    utterThis.text = text;
    utterThis.onend = () => {
      setPlayCircleRef("#F0F3F4");
      setPlayIconRef("#4F4F4F");
    };
    setPlayCircleRef("#515151");
    setPlayIconRef("white");
    synth.speak(utterThis);
  };

  const playAudio = () => {
    const audio = document.getElementById("audio");
    audio.onended = () => {
      setPlayCircleSelf("#F0F3F4");
      setPlayIconSelf("#4F4F4F");
    };
    setPlayCircleSelf("#515151");
    setPlayIconSelf("white");
    audio.play();
  };

  const handleNext = () => {
    setCurrent((current + 1) % examples.length);
  };

  const handlePrev = () => {
    let new_value = current - 1;
    if (new_value < 0) {
      new_value += examples.length;
    }
    setCurrent(new_value);
  };

  return (
    <Container id="demo">
      <Content>
        <VerticalContainer>
          <HeadingTitle>{title}</HeadingTitle>
          <HeadingDescription>{description}</HeadingDescription>
          <HorizontalContainer>
            <Tooltip title="Previous">
              <SvgPrevIcon onClick={handlePrev} />
            </Tooltip>
            <Image src={imageSrc} />
            <Tooltip title="Next">
              <SvgNextIcon onClick={handleNext} />
            </Tooltip>
          </HorizontalContainer>
          <TextToRead>{text}</TextToRead>
          <HorizontalContainer>
            <PhonemeViewer phonemes={target} examples={examplesPhonemes} />
            <PlayButton
              onClick={speakWord}
              circleColor={playCircleRef}
              iconColor={playIconRef}
            />
          </HorizontalContainer>
          {pred.length !== 0 ? (
            <HorizontalContainer>
              <PhonemeViewer phonemes={pred} examples={examplesPhonemes} />
              <PlayButton
                onClick={playAudio}
                circleColor={playCircleSelf}
                iconColor={playIconSelf}
              />
            </HorizontalContainer>
          ) : null}
          <RecordButton
            fillColor={recordColor}
            radius={radius}
            onClick={handleClick}
          />
          <AudioReactRecorder
            setRadius={setRadius}
            state={recordState}
            onStop={handleStop}
            backgroundColor="rgb(255,255,255)"
            canvasHeight={recorderHeight}
            canvasWidth="0"
          />
          {audioData ? (
            <Audio
              id="audio"
              // controls
              src={audioData ? audioData.url : null}
            />
          ) : null}
        </VerticalContainer>
      </Content>
      <DecoratorBlob1 />
      <DecoratorBlob2 />
    </Container>
  );
};


audio_recorder_streamlit/frontend/src/RecordButton.js
import React from "react"

export const RecordButton = ({
  onClick,
  fillColor = "white",
  color = "#6416FF",
  radius = "0",
}) => {
  return (
    <svg
      width="97"
      height="97"
      viewBox="0 0 97 97"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle cx="48.5" cy="48.5" r={radius} fill="#DCDCDC" />
      <circle cx="49" cy="49" r="33" fill={color} onClick={onClick} />
      <g clipPath="url(#clip0)">
        <path
          onClick={onClick}
          d="M49 57.2066C53.4997 57.2066 57.1474 53.5324 57.1474 49V35.3224C57.1474 30.7899 53.4997 27.1158 49 27.1158C44.5003 27.1158 40.8526 30.7899 40.8526 35.3224V49C40.8526 53.5324 44.5003 57.2066 49 57.2066ZM62.579 43.5289H61.2211C60.4708 43.5289 59.8632 44.141 59.8632 44.8967V49C59.8632 55.3943 54.39 60.5251 47.9145 59.8891C42.2708 59.3343 38.1369 54.224 38.1369 48.5127V44.8967C38.1369 44.141 37.5292 43.5289 36.779 43.5289H35.4211C34.6708 43.5289 34.0632 44.141 34.0632 44.8967V48.3298C34.0632 55.9927 39.4922 62.8238 46.9632 63.8616V66.7809H42.2105C41.4603 66.7809 40.8526 67.393 40.8526 68.1487V69.5164C40.8526 70.2721 41.4603 70.8842 42.2105 70.8842H55.7895C56.5397 70.8842 57.1474 70.2721 57.1474 69.5164V68.1487C57.1474 67.393 56.5397 66.7809 55.7895 66.7809H51.0369V63.8941C58.3109 62.8888 63.9369 56.5996 63.9369 49V44.8967C63.9369 44.141 63.3292 43.5289 62.579 43.5289Z"
          fill={fillColor}
        />
      </g>
      <defs>
        <clipPath id="clip0">
          <rect
            width="29.8737"
            height="43.7684"
            fill="white"
            transform="translate(34.0632 27.1158)"
          />
        </clipPath>
      </defs>
    </svg>
  )
}

export default RecordButton


audio-recorder/frontend/src/AudioRecorder.js
import React from "react";
import PropTypes from "prop-types"; // ES6

// export const AudioReactRecorder = ({ text }) => {
//   return <div className={styles.test}>BULLSWEET: {text}</div>
// }

export const RecordState = Object.freeze({
  START: "start",
  PAUSE: "pause",
  STOP: "stop",
  NONE: "none",
});

export default class AudioReactRecorder extends React.Component {
  //0 - constructor
  constructor(props) {
    super(props);

    this.canvasRef = React.createRef();
  }

  //TODO: add the props definitions
  static propTypes = {
    state: PropTypes.string,
    type: PropTypes.string.isRequired,
    backgroundColor: PropTypes.string,
    foregroundColor: PropTypes.string,
    canvasWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    canvasHeight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

    //method calls
    onStop: PropTypes.func,
  };
  static defaultProps = {
    state: RecordState.NONE,
    type: "audio/wav",
    backgroundColor: "rgb(200, 200, 200)",
    foregroundColor: "rgb(0, 0, 0)",
    canvasWidth: 500,
    canvasHeight: 150,
    pause_threshold: 0.8,
    phrase_threshold: 0.3,
  };

  //2 - mount
  componentDidMount() {
    this.init();
  }

  componentDidUpdate(prevProps, prevState) {
    const { state } = this.props;

    this.checkState(prevProps.state, state);
  }

  checkState(previousState) {
    switch (previousState) {
      case RecordState.START:
        this.doIfState(RecordState.PAUSE, this.pause);
        this.doIfState(RecordState.STOP, this.stop);
        break;
      case RecordState.PAUSE:
        this.doIfState(RecordState.START, this.resume);
        this.doIfState(RecordState.STOP, this.stop);
        break;
      case RecordState.STOP:
        this.doIfState(RecordState.START, this.start);
        break;
      default:
        this.doIfState(RecordState.START, this.start);
        break;
    }
  }

  doIfState(state, cb) {
    if (this.props.state === state) {
      cb && cb();
    }
  }

  //TODO: destroy request animation frame
  componentWillUnmount() {}

  //TODO: change to state some conditionals
  init = async () => {
    this.leftchannel = [];
    this.rightchannel = [];
    this.recorder = null;
    this.recording = false;
    this.recordingLength = 0;
    this.volume = null;
    this.audioInput = null;
    this.sampleRate = null;
    this.AudioContext = window.AudioContext || window.webkitAudioContext;
    this.context = null;
    this.analyser = null;
    this.canvas = this.canvasRef.current;
    this.canvasCtx = this.canvas.getContext("2d");
    this.stream = null;
    this.tested = false;
    this.phrase_buffer_count = null;
    this.pause_buffer_count = null;
    this.energy_threshold = 0.01;
    this.pause_count = null;
    this.stage = null;

    navigator.getUserMedia =
      navigator.getUserMedia ||
      navigator.webkitGetUserMedia ||
      navigator.mozGetUserMedia;
  };

  //get mic stream
  getStream = (constraints) => {
    if (!constraints) {
      constraints = { audio: true, video: false };
    }

    return navigator.mediaDevices.getUserMedia(constraints);
  };

  setUpRecording = () => {
    this.context = new this.AudioContext();
    this.sampleRate = this.context.sampleRate;

    // create buffer states counts
    let bufferSize = 2048;
    let seconds_per_buffer = bufferSize / this.sampleRate;
    this.phrase_buffer_count = Math.ceil(
      this.props.phrase_threshold / seconds_per_buffer
    );
    this.pause_buffer_count = Math.ceil(
      this.props.pause_threshold / seconds_per_buffer
    );
    this.pause_count = 0;
    this.stage = "start";

    // creates a gain node
    this.volume = this.context.createGain();

    // creates an audio node from teh microphone incoming stream
    this.audioInput = this.context.createMediaStreamSource(this.stream);

    // Create analyser
    this.analyser = this.context.createAnalyser();

    // connect audio input to the analyser
    this.audioInput.connect(this.analyser);

    // connect analyser to the volume control
    // analyser.connect(volume);

    this.recorder = this.context.createScriptProcessor(bufferSize, 2, 2);

    // we connect the volume control to the processor
    // volume.connect(recorder);

    this.analyser.connect(this.recorder);

    // finally connect the processor to the output
    this.recorder.connect(this.context.destination);

    const self = this;
    this.recorder.onaudioprocess = function (e) {
      // Check
      if (!self.recording) return;
      // Do something with the data, i.e Convert this to WAV
      let left = e.inputBuffer.getChannelData(0);
      let right = e.inputBuffer.getChannelData(1);
      if (!self.tested) {
        self.tested = true;
        // if this reduces to 0 we are not getting any sound
        if (!left.reduce((a, b) => a + b)) {
          console.log("Error: There seems to be an issue with your Mic");
          // clean up;
          self.stop();
          self.stream.getTracks().forEach(function (track) {
            track.stop();
          });
          self.context.close();
        }
      }
      // Check energy level
      let energy = Math.sqrt(
        left.map((x) => x * x).reduce((a, b) => a + b) / left.length
      );
      if (self.stage === "start" && energy > self.energy_threshold) {
        self.stage = "speaking";
      } else if (self.stage === "speaking") {
        if (energy > self.energy_threshold) {
          self.pause_count = 0;
        } else {
          self.pause_count += 1;
          if (self.pause_count > self.pause_buffer_count) {
            self.stop();
          }
        }
      }
      let radius = 33.0 + Math.sqrt(1000.0 * energy);
      self.props.setRadius(radius.toString());

      // we clone the samples
      self.leftchannel.push(new Float32Array(left));
      self.rightchannel.push(new Float32Array(right));
      self.recordingLength += bufferSize;
    };
    this.visualize();
  };

  mergeBuffers = (channelBuffer, recordingLength) => {
    let result = new Float32Array(recordingLength);
    let offset = 0;
    let lng = channelBuffer.length;
    for (let i = 0; i < lng; i++) {
      let buffer = channelBuffer[i];
      result.set(buffer, offset);
      offset += buffer.length;
    }
    return result;
  };

  interleave = (leftChannel, rightChannel) => {
    let length = leftChannel.length + rightChannel.length;
    let result = new Float32Array(length);

    let inputIndex = 0;

    for (let index = 0; index < length; ) {
      result[index++] = leftChannel[inputIndex];
      result[index++] = rightChannel[inputIndex];
      inputIndex++;
    }
    return result;
  };

  writeUTFBytes = (view, offset, string) => {
    let lng = string.length;
    for (let i = 0; i < lng; i++) {
      view.setUint8(offset + i, string.charCodeAt(i));
    }
  };

  visualize = () => {
    const { backgroundColor, foregroundColor } = this.props;

    this.WIDTH = this.canvas.width;
    this.HEIGHT = this.canvas.height;
    this.CENTERX = this.canvas.width / 2;
    this.CENTERY = this.canvas.height / 2;

    if (!this.analyser) return;

    this.analyser.fftSize = 2048;
    const bufferLength = this.analyser.fftSize;
    const dataArray = new Uint8Array(bufferLength);

    this.canvasCtx.clearRect(0, 0, this.WIDTH, this.HEIGHT);

    //reference this using self
    let self = this;
    const draw = function () {
      self.drawVisual = requestAnimationFrame(draw);

      self.analyser.getByteTimeDomainData(dataArray);

      self.canvasCtx.fillStyle = backgroundColor;
      self.canvasCtx.fillRect(0, 0, self.WIDTH, self.HEIGHT);

      self.canvasCtx.lineWidth = 2;
      self.canvasCtx.strokeStyle = foregroundColor;

      self.canvasCtx.beginPath();

      var sliceWidth = (self.WIDTH * 1.0) / bufferLength;
      var x = 0;

      for (var i = 0; i < bufferLength; i++) {
        var v = dataArray[i] / 128.0;
        var y = (v * self.HEIGHT) / 2;

        if (i === 0) {
          self.canvasCtx.moveTo(x, y);
        } else {
          self.canvasCtx.lineTo(x, y);
        }

        x += sliceWidth;
      }

      self.canvasCtx.lineTo(self.canvas.width, self.canvas.height / 2);
      self.canvasCtx.stroke();
    };

    draw();
  };

  setupMic = async () => {
    //TODO: only get stream after clicking start
    try {
      window.stream = this.stream = await this.getStream();
      //TODO: on got stream
    } catch (err) {
      //TODO: error getting stream
      console.log("Error: Issue getting mic", err);
    }

    this.setUpRecording();
  };

  start = async () => {
    await this.setupMic();

    this.recording = true;
    // reset the buffers for the new recording
    this.leftchannel.length = this.rightchannel.length = 0;
    this.recordingLength = 0;
  };

  stop = () => {
    const { onStop, type } = this.props;

    this.recording = false;
    this.closeMic();

    // we flat the left and right channels down
    this.leftBuffer = this.mergeBuffers(this.leftchannel, this.recordingLength);
    this.rightBuffer = this.mergeBuffers(
      this.rightchannel,
      this.recordingLength
    );
    // we interleave both channels together
    let interleaved = this.interleave(this.leftBuffer, this.rightBuffer);

    ///////////// WAV Encode /////////////////
    // from http://typedarray.org/from-microphone-to-wav-with-getusermedia-and-web-audio/
    //

    // we create our wav file
    let buffer = new ArrayBuffer(44 + interleaved.length * 2);
    let view = new DataView(buffer);

    // RIFF chunk descriptor
    this.writeUTFBytes(view, 0, "RIFF");
    view.setUint32(4, 44 + interleaved.length * 2, true);
    this.writeUTFBytes(view, 8, "WAVE");
    // FMT sub-chunk
    this.writeUTFBytes(view, 12, "fmt ");
    view.setUint32(16, 16, true);
    view.setUint16(20, 1, true);
    // stereo (2 channels)
    view.setUint16(22, 2, true);
    view.setUint32(24, this.sampleRate, true);
    view.setUint32(28, this.sampleRate * 4, true);
    view.setUint16(32, 4, true);
    view.setUint16(34, 16, true);
    // data sub-chunk
    this.writeUTFBytes(view, 36, "data");
    view.setUint32(40, interleaved.length * 2, true);

    // write the PCM samples
    let lng = interleaved.length;
    let index = 44;
    let volume = 1;
    for (let i = 0; i < lng; i++) {
      view.setInt16(index, interleaved[i] * (0x7fff * volume), true);
      index += 2;
    }

    // our final binary blob
    const blob = new Blob([view], { type: type });
    const audioUrl = URL.createObjectURL(blob);

    onStop &&
      onStop({
        blob: blob,
        url: audioUrl,
        type,
      });
  };

  pause = () => {
    this.recording = false;
    this.closeMic();
  };

  resume = () => {
    this.setupMic();
    this.recording = true;
  };

  closeMic = () => {
    this.stream.getAudioTracks().forEach((track) => {
      track.stop();
    });
    this.audioInput.disconnect(0);
    this.analyser.disconnect(0);
    this.recorder.disconnect(0);
  };

  //1 - render
  render() {
    const { canvasWidth, canvasHeight } = this.props;

    return (
      <div className="audio-react-recorder">
        <canvas
          ref={this.canvasRef}
          width={canvasWidth}
          height={canvasHeight}
          className="audio-react-recorder__canvas"
        ></canvas>
      </div>
    );
  }
}
